lib/ansible/module_utils/basic.py

```diff
@@ -652,6 +653,9 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
         self.no_log_values = set()
         # Use the argspec to determine which args are no_log
         for arg_name, arg_opts in self.argument_spec.items():
+            if arg_opts.get('deprecated', False):
+                self._deprecated_params.append(arg_name)
```

Comment:
why not just create deprecated warnings here and append directly in exit/fail_jsons?